File: android/app/src/main/java/com/audiotracker/menteenotes/MenteeNotesDb.java
package com.audiotracker.menteenotes;

import android.database.Cursor;
import android.util.Log;

import io.requery.android.database.sqlite.SQLiteDatabase;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Writes a mentee's synced rows into this device's database.
 *
 * It also creates them — see ensureTables. menteeNotesDB.js reads through the
 * same tables but never writes one, so the schema lives with the only thing
 * that does. Rows are someone else's writing: read-only to the user, never
 * part of their own search (mentee_notes is a plain table, not the fts5 one),
 * and dropped whole when a mentorship ends.
 *
 * Every value in a backup row arrives as a string, because the writer reads
 * them with Cursor.getString. The two numeric columns are coerced here rather
 * than left to surface as "1234" somewhere that wants a number.
 */
public final class MenteeNotesDb {

    private static final String TAG = "MenteeNotes";

    private MenteeNotesDb() {}

    /**
     * Creates the mentee tables if this device has never synced one.
     *
     * Called at the top of a sync rather than at login, so an account that
     * never mentors anybody never carries them — which is most accounts, and
     * about 60KB each. Every statement is IF NOT EXISTS, so the usual case is
     * seven no-ops on an open database.
     *
     * The reader tolerates their absence, because between a mentorship being
     * approved and the first sync finishing that is the true state.
     */
    public static void ensureTables(SQLiteDatabase db) {
        for (String sql : MenteeNotesSql.CREATE_SCHEMA) {
            db.execSQL(sql);
        }
    }

    /* ---------------------------------- */
    /* Sync state                          */
    /* ---------------------------------- */

    public static final class SyncState {
        public String folderId;
        public String imagesFolderId;
        public List<Coverage.Range> covered = new ArrayList<>();
        public long lastSyncAt;
    }

    public static SyncState loadSyncState(SQLiteDatabase db, String menteeId) {
        SyncState state = new SyncState();
        Cursor c = null;
        try {
            c = db.rawQuery(MenteeNotesSql.SELECT_SYNC_STATE, new String[]{menteeId});
            if (c.moveToFirst()) {
                state.folderId = c.getString(0);
                state.imagesFolderId = c.getString(1);
                state.covered = Coverage.parse(c.getString(2));
                state.lastSyncAt = c.getLong(3);
            }
        } catch (Exception e) {
            Log.e(TAG, "loadSyncState failed", e);
        } finally {
            if (c != null) c.close();
        }
        return state;
    }

    public static void saveSyncState(SQLiteDatabase db, String menteeId, SyncState state) {
        db.execSQL(MenteeNotesSql.UPSERT_SYNC_STATE, new Object[]{
                menteeId,
                state.folderId,
                state.imagesFolderId,
                Coverage.serialize(state.covered),
                state.lastSyncAt,
        });
    }

    /**
     * Forgets everything read for this mentee.
     *
     * For a folder id that has changed — a reinstall backs up somewhere new,
     * and coverage recorded against the old one describes a history that no
     * longer exists — and for a mentorship that has ended.
     */
    public static void clearMentee(SQLiteDatabase db, String menteeId) {
        for (String table : MenteeNotesSql.OWNED_TABLES) {
            try {
                db.execSQL(MenteeNotesSql.deleteFor(table), new Object[]{menteeId});
            } catch (Exception e) {
                Log.e(TAG, "Could not clear " + table, e);
            }
        }
    }

    /* ---------------------------------- */
    /* Applying a downloaded level file    */
    /* ---------------------------------- */

    /**
     * Everything one level file contributes, in one transaction.
     *
     * Items before notes: a note is stored whether or not its media is known,
     * but the title is what makes it mean anything, and when both are in the
     * same file there is no reason for the note to land first and be untitled
     * for a moment.
     *
     * Returns how many notes were written.
     */
    public static int applyLevelFile(SQLiteDatabase db, String menteeId, JSONObject data) {
        int notes = 0;
        db.beginTransaction();
        try {
            applyItems(db, menteeId, data.optJSONArray("items"));
            applyYoutubeMeta(db, menteeId, data.optJSONArray("youtube_meta"));
            applyDriveCopies(db, menteeId, data.optJSONArray("shared_drive_copies"));
            notes = applyNotes(db, menteeId, data.optJSONArray("notes"));
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e(TAG, "applyLevelFile failed", e);
        } finally {
            db.endTransaction();
        }
        return notes;
    }

    private static int applyNotes(SQLiteDatabase db, String menteeId, JSONArray rows) {
        if (rows == null) return 0;
        int applied = 0;

        for (int i = 0; i < rows.length(); i++) {
            JSONObject row = rows.optJSONObject(i);
            if (row == null || row.isNull("rowid")) continue;

            try {
                db.execSQL(MenteeNotesSql.UPSERT_NOTE, new Object[]{
                        menteeId,
                        row.optLong("rowid"),
                        str(row, "source_id"),
                        str(row, "source_type"),
                        str(row, "title"),
                        str(row, "content"),
                        // Backups written before entity decoding carry raw
                        // "&nbsp;" here. Left as found: the JS reader decodes
                        // this user's own notes the same way on the way in,
                        // and doing it in two places with two implementations
                        // is how they come to disagree.
                        str(row, "text_content"),
                        str(row, "created_at"),
                        str(row, "updated_at"),
                        str(row, "deleted_at"),
                });
                applied++;
            } catch (Exception e) {
                Log.e(TAG, "note upsert failed", e);
            }
        }
        return applied;
    }

    private static void applyItems(SQLiteDatabase db, String menteeId, JSONArray rows) {
        if (rows == null) return;

        for (int i = 0; i < rows.length(); i++) {
            JSONObject row = rows.optJSONObject(i);
            if (row == null) continue;
            String sourceId = str(row, "source_id");
            String type = str(row, "type");
            if (sourceId == null || type == null) continue;

            try {
                db.execSQL(MenteeNotesSql.UPSERT_ITEM, new Object[]{
                        menteeId,
                        sourceId,
                        type,
                        num(row, "id"),
                        str(row, "title"),
                        str(row, "mimeType"),
                        num(row, "duration"),
                        // Their path, useless as a path here — kept because
                        // the reader rebuilds an iskcon url from it, and that
                        // url is what makes the file streamable on this
                        // device.
                        str(row, "file_path"),
                        str(row, "updated_at"),
                });
            } catch (Exception e) {
                Log.e(TAG, "item upsert failed", e);
            }
        }
    }

    // youtube_meta and shared_drive_copies are keyed by the mentee's own
    // items.id, which is why mentee_items carries remote_id. Two writers into
    // one row, each touching only its own columns, so neither clobbers the
    // other and the order they arrive in does not matter.
    private static void applyYoutubeMeta(SQLiteDatabase db, String menteeId, JSONArray rows) {
        if (rows == null) return;
        for (int i = 0; i < rows.length(); i++) {
            JSONObject row = rows.optJSONObject(i);
            Long remoteId = row == null ? null : num(row, "item_id");
            if (remoteId == null) continue;
            try {
                db.execSQL(MenteeNotesSql.UPSERT_YOUTUBE_META, new Object[]{
                        menteeId, remoteId, str(row, "channel_title"), str(row, "thumbnail"),
                });
            } catch (Exception e) {
                Log.e(TAG, "youtube meta upsert failed", e);
            }
        }
    }

    private static void applyDriveCopies(SQLiteDatabase db, String menteeId, JSONArray rows) {
        if (rows == null) return;
        for (int i = 0; i < rows.length(); i++) {
            JSONObject row = rows.optJSONObject(i);
            Long remoteId = row == null ? null : num(row, "item_id");
            if (remoteId == null) continue;
            try {
                db.execSQL(MenteeNotesSql.UPSERT_DRIVE_COPY, new Object[]{
                        menteeId, remoteId, str(row, "drive_file_id"),
                });
            } catch (Exception e) {
                Log.e(TAG, "drive copy upsert failed", e);
            }
        }
    }

    /* ---------------------------------- */
    /* Images                              */
    /* ---------------------------------- */

    public static void applyImages(SQLiteDatabase db, String menteeId, JSONArray rows) {
        if (rows == null) return;
        db.beginTransaction();
        try {
            for (int i = 0; i < rows.length(); i++) {
                JSONObject row = rows.optJSONObject(i);
                Long imageId = row == null ? null : num(row, "id");
                if (imageId == null) continue;
                db.execSQL(MenteeNotesSql.UPSERT_IMAGE, new Object[]{
                        menteeId,
                        imageId,
                        num(row, "note_rowid"),
                        str(row, "image_data"),
                        str(row, "updated_at"),
                        str(row, "deleted_at"),
                });
            }
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e(TAG, "applyImages failed", e);
        } finally {
            db.endTransaction();
        }
    }

    public static Set<String> loadSeenFiles(SQLiteDatabase db, String menteeId) {
        Set<String> seen = new HashSet<>();
        Cursor c = null;
        try {
            c = db.rawQuery(MenteeNotesSql.SELECT_SEEN_FILES, new String[]{menteeId});
            while (c.moveToNext()) seen.add(c.getString(0));
        } catch (Exception e) {
            Log.e(TAG, "loadSeenFiles failed", e);
        } finally {
            if (c != null) c.close();
        }
        return seen;
    }

    public static void markFileSeen(SQLiteDatabase db, String menteeId, String name) {
        try {
            db.execSQL(MenteeNotesSql.INSERT_SEEN_FILE, new Object[]{menteeId, name});
        } catch (Exception e) {
            Log.e(TAG, "markFileSeen failed", e);
        }
    }

    /* ---------------------------------- */
    /* Field readers                       */
    /* ---------------------------------- */

    // JSONObject.NULL is not null, and optString turns it into the four
    // characters "null" — which would be written into the column as if it were
    // a title.
    private static String str(JSONObject row, String key) {
        if (row.isNull(key)) return null;
        return row.optString(key, null);
    }

    private static Long num(JSONObject row, String key) {
        if (row.isNull(key)) return null;
        try {
            return Long.valueOf(row.getString(key));
        } catch (Exception e) {
            return null;
        }
    }
}

File: src/database/menteeNotesDB.js
// menteeNotesDB.js
//
// Reading the notes a mentee's Drive backup was synced into.
//
// Reading only. The writing is native — see android/.../menteenotes — because
// it has to happen on a phone nobody has opened, which no JS can promise.
// That side also creates the tables, on its first sync, so a user with no
// mentees never carries them: the query below can find nothing to read from,
// and treats that as the empty answer it is.
//
// Separate tables from this user's own, because they are someone else's
// writing: read-only here, and never part of this user's search. mentee_notes
// is a plain table, deliberately, while `notes` is the fts5 index behind their
// searches — a mentee's words landing in that would answer them.

import {getDb} from './database';
import {noteRef} from '../notes/noteRef';
import {getShareLink} from '../Linking/utils/shareLink';

/* ---------------------------------- */
/* Reading, for the UI                 */
/* ---------------------------------- */

export const fetchMenteeNotes = ({
  menteeId,
  sourceId = null,
  date = null,
  searchQuery = null,
  offset = 0,
  limit = 20,
} = {}) => {
  const fastdb = getDb();

  return new Promise((resolve, reject) => {
    // A deleted note arrives as a tombstone with its text blanked, not as a
    // missing row — that is how the mentee's own delete propagates.
    //
    // relatedItem is NULL rather than an object of nulls when nothing matches.
    // NoteItem renders the source line behind `item.relatedItem &&`, so a null
    // hides it; an empty object is truthy and would print "youtube_video:
    // null". The miss is normal — a notebook note has no item, and neither
    // does one whose item has not come down yet.
    let query = `
      SELECT n.note_rowid AS rowid,
             n.source_id,
             n.source_type,
             n.title AS noteTitle,
             n.content,
             n.text_content,
             n.created_at,
             n.updated_at,
             'note' AS type,
             n.mentee_id AS ownerId,
             CASE WHEN mi.source_id IS NOT NULL
               THEN json_object('source_id',     mi.source_id,
                                'type',          mi.type,
                                'title',         mi.title,
                                'mimeType',      mi.mimeType,
                                'duration',      mi.duration,
                                'file_path',     mi.file_path,
                                'channel_title', mm.channel_title,
                                'thumbnail',     mm.thumbnail,
                                'drive_file_id', mm.drive_file_id)
               ELSE NULL
             END AS relatedItem
        FROM mentee_notes n
        LEFT JOIN mentee_items mi
               ON mi.mentee_id = n.mentee_id
              AND mi.source_id = n.source_id
              AND mi.type      = n.source_type
        LEFT JOIN mentee_item_meta mm
               ON mm.mentee_id = mi.mentee_id
              AND mm.remote_id = mi.remote_id
       WHERE n.mentee_id = ? AND n.deleted_at IS NULL`;
    const params = [menteeId];

    if (sourceId) {
      query += ' AND n.source_id = ?';
      params.push(sourceId);
    }
    // Matched on created_at, the same column the user's own day report uses,
    // so a note edited later still belongs to the day it was written.
    if (date) {
      query += ' AND DATE(n.created_at) = ?';
      params.push(date);
    }
    if (searchQuery) {
      query += ' AND (n.title LIKE ? OR n.text_content LIKE ?)';
      params.push(`%${searchQuery}%`, `%${searchQuery}%`);
    }

    query += ' ORDER BY n.created_at DESC LIMIT ? OFFSET ?;';
    params.push(limit, offset);

    fastdb.transaction(tx => {
      tx.executeSql(
        query,
        params,
        (_, {rows}) => {
          const out = [];
          for (let i = 0; i < rows.length; i++) {
            const note = rows.item(i);
            // Self-describing from birth: a bare rowid would be looked up in
            // this user's own notes table by anything downstream, and 47 is a
            // real note in both. Nothing between here and the read has to
            // carry a second value.
            note.rowid = noteRef(note.rowid, note.ownerId);
            delete note.ownerId;
            try {
              note.relatedItem = note.relatedItem
                ? JSON.parse(note.relatedItem)
                : null;
            } catch {
              note.relatedItem = null;
            }

            // Shaped here, once, rather than at the point of a tap: the row
            // that leaves this function is the row the player takes, so
            // nothing downstream needs a branch for whose note it is.
            //
            // file_path is the field the player reads for a streamable url,
            // and an iskcon file is the only type that can supply one from
            // what we hold — its source_id is the path on the site, so
            // getShareLink rebuilds it. Drive and device files resolve
            // through resolvePlaybackPath instead, off source_id and
            // drive_file_id, and a copy downloaded onto the mentee's phone is
            // not something this device could open anyway.
            if (note.relatedItem?.type === 'iskcon_file') {
              note.relatedItem.file_path = getShareLink(note.relatedItem);
            } else if (note.relatedItem) {
              note.relatedItem.file_path = null;
            }
            out.push(note);
          }
          resolve(out);
        },
        (_, error) => {
          // No tables yet. The native sync creates them on its first run, so
          // this is the honest state of a mentorship that was approved
          // minutes ago — and the permanent state of everyone who never
          // mentors anybody. Either way the answer is that there are no
          // notes, not that something went wrong.
          if (/no such table/i.test(error?.message ?? '')) {
            resolve([]);
            return;
          }
          reject(error);
        },
      );
    });
  });
};

/**
 * The media row a mentee's note was taken against, or null.
 *
 * The same shape getItemBySourceId returns for this user's own items, so the
 * one caller that needs both — describeNoteMedia, building the media half of
 * a shared note — differs only in which of the two it asks.
 *
 * channel_title/thumbnail/drive_file_id come off mentee_item_meta rather than
 * the item, exactly as youtube_meta and shared_drive_copies do on the user's
 * own side; flattened onto the row here because that is how the fields are
 * read downstream.
 */
export const getMenteeItemBySourceId = ({menteeId, sourceId, type}) => {
  const fastdb = getDb();

  return new Promise((resolve, reject) => {
    fastdb.transaction(tx => {
      tx.executeSql(
        `SELECT mi.source_id,
                mi.type,
                mi.title,
                mi.mimeType,
                mi.duration,
                mi.file_path,
                mm.channel_title,
                mm.thumbnail,
                mm.drive_file_id
           FROM mentee_items mi
           LEFT JOIN mentee_item_meta mm
                  ON mm.mentee_id = mi.mentee_id
                 AND mm.remote_id = mi.remote_id
          WHERE mi.mentee_id = ? AND mi.source_id = ? AND mi.type = ?;`,
        [menteeId, sourceId, type],
        (_, {rows: {_array}}) => resolve(_array[0] || null),
        (_, error) => {
          // No tables yet — the same honest empty answer fetchMenteeNotes
          // gives. A note with no item behind it is an ordinary outcome here,
          // so the caller already has a branch for null.
          if (/no such table/i.test(error?.message ?? '')) {
            resolve(null);
            return;
          }
          reject(error);
          return false;
        },
      );
    });
  });
};

/**
 * How many notes this mentee has written against each item, keyed by source_id.
 *
 * One grouped query rather than a count per row: a list is twenty rows deep and
 * every one of them would otherwise open its own read for a number that comes
 * out of a single GROUP BY.
 *
 * Keyed on source_id alone, which is safe because an assignment gives both
 * sides the same one - see ingestDeviceAssignment, where the mentor's id is
 * deliberately what the mentee's row is built with. Notebook notes are left out:
 * their source_id is a notebook rowid, which belongs to a different numbering
 * altogether and has no row on a mentor's screen to sit under.
 *
 * Tombstones are excluded the same way fetchMenteeNotes excludes them - a
 * deleted note arrives as a blanked row, not a missing one, so counting rows
 * without this would keep counting notes the mentee has thrown away.
 */
export const countMenteeNotesBySource = menteeId => {
  const fastdb = getDb();

  return new Promise((resolve, reject) => {
    if (!menteeId) {
      resolve({});
      return;
    }

    fastdb.transaction(tx => {
      tx.executeSql(
        `SELECT source_id, COUNT(*) AS noteCount
           FROM mentee_notes
          WHERE mentee_id = ?
            AND deleted_at IS NULL
            AND source_id IS NOT NULL
            AND source_type <> 'notebook'
          GROUP BY source_id;`,
        [menteeId],
        (_, {rows}) => {
          const counts = {};
          for (let i = 0; i < rows.length; i++) {
            const row = rows.item(i);
            counts[String(row.source_id)] = row.noteCount;
          }
          resolve(counts);
        },
        (_, error) => {
          // No tables yet - the same honest empty answer the reads above give.
          if (/no such table/i.test(error?.message ?? '')) {
            resolve({});
            return;
          }
          reject(error);
          return false;
        },
      );
    });
  });
};
